### PR TITLE
Recover repeated zero-byte tool input restarts

### PR DESCRIPTION
**Pull request by steve8708**
## Summary
- recover no-progress agent turns when the model repeatedly emits fresh zero-byte tool-input deltas for the same action
- preserve support for genuinely large outputs by resetting the restart state when positive tool-input bytes stream
- add a regression test for repeated zero-byte edit-design deltas with fresh IDs

## Validation
- corepack pnpm --filter @agent-native/core exec vitest run src/agent/production-agent.spec.ts src/agent/run-loop-with-resume.spec.ts src/agent/tool-call-journal.spec.ts
- corepack pnpm --filter @agent-native/core typecheck
- corepack pnpm prep

## Production Evidence
- Fresh production Design QA run MVv1ghNInAn_VxRBq2sMN created three variants successfully on main@a234670.
- Selected-variant build then stalled on repeated zero-byte edit-design prep events with fresh IDs and no tool_start/tool_done, matching the fixed path here.

### CODE REVIEW

File: packages/core/src/agent/production-agent.ts [hunk 1]
```diff
@@ -2983,10 +2984,24 @@ export async function runAgentLoop(opts: {
                 trackActiveToolInput(key, toolName, progressBytes);
                 if (progressBytes > 0) {
                   resetZeroByteToolInputRestart(toolName);
+                } else if (!hadByteRecord) {
+                  startedZeroByteInput = true;
```
[the remark this answers is not in the document]
Comment (steve8708) [reply packages/core/src/agent/production-agent.ts:2987]: Fixed in c0120f9: positive-byte progress on a replacement stream now retires older zero-byte active inputs for the same tool before the stall watchdog checks them again. I also added a regression where an abandoned zero-byte id is followed by a fresh edit-design id streaming bytes past the original 90s boundary; it now completes without auto_continue.

File: packages/core/src/agent/production-agent.ts [hunk 2]
```diff
@@ -2874,11 +2874,21 @@ export async function runAgentLoop(opts: {
...
           const previous = activeToolInputs.get(key);
+          const resolvedToolName = toolName ?? previous?.toolName;
+          if (bytes > 0 && resolvedToolName) {
+            for (const [activeKey, active] of activeToolInputs) {
+              if (
+                activeKey !== key &&
+                active.toolName === resolvedToolName &&
+                active.bytes === 0
+              ) {
+                activeToolInputs.delete(activeKey);
```
[the remark this answers is not in the document]
Comment (steve8708) [reply packages/core/src/agent/production-agent.ts]: Fixed in 2a8aef4: stale zero-byte retirement is now limited to serial mutating actions. Read-only, parallel-safe, and unknown tools keep same-action inputs tracked independently, and I added a regression where a parallel-safe same-action sibling streams bytes while the older input still triggers no_progress.

File: packages/core/src/agent/production-agent.ts [hunk 3]
```diff
@@ -2867,18 +2867,39 @@ export async function runAgentLoop(opts: {
           }
           return false;
         };
+        const allowsParallelInputPreparation = (toolName?: string) => {
+          if (!toolName) return true;
+          const entry = actions[toolName];
+          return (
+            !entry || entry.readOnly === true || entry.parallelSafe === true
```
[the remark this answers is not in the document]
Comment (steve8708) [reply packages/core/src/agent/production-agent.ts]: Fixed in 513f46e: stale cleanup is no longer based on action readOnly/parallelSafe. Active input records now remember whether they were created implicitly from a delta without a start event, and positive-byte replacement streams only retire those implicit zero-byte records. This keeps normal start-event parallel inputs tracked while covering read-only and mutating fresh-ID recovery. Added read-only regression coverage too.

File: packages/core/src/agent/production-agent.ts [hunk 4]
```diff
@@ -2871,17 +2872,30 @@ export async function runAgentLoop(opts: {
...
+          const resolvedToolName = toolName ?? previous?.toolName;
+          if (bytes > 0 && resolvedToolName) {
+            for (const [activeKey, active] of activeToolInputs) {
+              if (
+                activeKey !== key &&
+                active.toolName === resolvedToolName &&
+                active.bytes === 0 &&
+                active.startedFromDelta
+              ) {
+                activeToolInputs.delete(activeKey);
```
[the remark this answers is not in the document]
Comment (steve8708) [reply packages/core/src/agent/production-agent.ts]: Fixed in 933049d: I removed the same-tool sibling cleanup entirely. The action-preparation watchdog now keys off the most recent tool-input progress across all active preparations, so a healthy replacement stream or delta-only sibling keeps the run alive, and true silence still checkpoints with no_progress. Added regression coverage for delta-only same-action siblings plus updated multi-input stall tests to wait for 91s of no input progress before recovering.
